File: flutter_client/lib/modules/auth/views/server_list/udp_broadcast_listener.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';

import 'package:flutter/foundation.dart';

/// UDP广播监听工具类
class UdpBroadcastListener {
  static const int _broadcastPort = 8888;
  RawDatagramSocket? _socket;
  bool _isListening = false;

  /// 服务器发现回调函数
  Function(Map<String, dynamic>)? _onServerDiscovered;

  /// 设置服务器发现回调
  void setOnServerDiscovered(Function(Map<String, dynamic>) callback) {
    _onServerDiscovered = callback;
  }

  /// 开始监听UDP广播
  Future<void> startListening() async {
    if (kIsWeb) {
      print('UDP广播监听在Web平台上不支持');
      return;
    }
    if (_isListening) {
      print('UDP监听器已经在运行中');
      return;
    }

    try {
      // 创建UDP socket
      _socket = await RawDatagramSocket.bind(
        InternetAddress.anyIPv4,
        _broadcastPort,
        reuseAddress: true,
      );

      // 监听广播消息
      _socket!.listen(
        (RawSocketEvent event) {
          if (event == RawSocketEvent.read) {
            final datagram = _socket!.receive();
            if (datagram != null) {
              final data = datagram.data;
              final sender = datagram.address;

              // 将接收到的数据转换为字符串
              final message = String.fromCharCodes(data);
              // print(message);
              try {
                // 解析JSON消息
                final Map<String, dynamic> serverInfo = json.decode(message);

                // 验证必需字段
                if (serverInfo.containsKey('service') &&
                    serverInfo['service'] == 'nascab-pro-service' &&
                    serverInfo.containsKey('host') &&
                    serverInfo.containsKey('port')) {
                  // 添加发送者IP信息
                  serverInfo['discovered_ip'] = sender.address;

                  // 调用服务器发现回调
                  if (_onServerDiscovered != null) {
                    _onServerDiscovered!(serverInfo);
                  }

                  // print(
                  //   '📡 发现服务器: ${serverInfo['hostname']} (${serverInfo['host']}:${serverInfo['port']})',
                  // );
                }
              } catch (e) {
                print('❌ 解析UDP消息失败: $e');
              }
            }
          }
        },
        onError: (error) {
          print('❌ UDP监听错误: $error');
        },
        onDone: () {
          print('✅ UDP监听完成');
          _isListening = false;
        },
      );

      _isListening = true;
      print('🎯 UDP广播监听器已启动，监听端口: $_broadcastPort');
      print('等待接收广播消息...');
    } catch (e) {
      print('❌ 启动UDP监听失败: $e');
      _isListening = false;
    }
  }

  /// 停止监听UDP广播
  Future<void> stopListening() async {
    stopListeningSync();
  }

  /// 同步关闭 socket，供 GetX [onClose] 等必须立即释放端口的场景使用。
  void stopListeningSync() {
    if (!_isListening && _socket == null) {
      return;
    }
    try {
      _socket?.close();
      _socket = null;
      _isListening = false;
      print('🛑 UDP广播监听器已停止');
    } catch (e) {
      print('❌ 停止UDP监听失败: $e');
      _socket = null;
      _isListening = false;
    }
  }

  /// 检查是否正在监听
  bool get isListening => _isListening;

  /// 获取监听端口
  int get listeningPort => _broadcastPort;

  /// 析构函数，确保资源被正确释放
  void dispose() {
    stopListeningSync();
  }
}

File: tv_apple/NasCabTV/Core/Network/UdpBroadcastListener.swift
import Foundation

/// UDP 广播监听：监听端口 8888，解析 NasCab 服务发现报文，与 Flutter 端 UdpBroadcastListener 协议一致
final class UdpBroadcastListener {
    static let broadcastPort: UInt16 = 8888

    typealias ServerDiscoveredHandler = ([String: Any]) -> Void

    private var socket: Int32 = -1
    private var receiveSource: DispatchSourceRead?
    private let queue = DispatchQueue(label: "UdpBroadcastListener.receive")
    private var onServerDiscovered: ServerDiscoveredHandler?

    var isListening: Bool { socket >= 0 }

    func setOnServerDiscovered(_ handler: @escaping ServerDiscoveredHandler) {
        onServerDiscovered = handler
    }

    func startListening() {
        guard socket < 0 else { return }

        let fd = Darwin.socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
        guard fd >= 0 else {
            print("[UDP] socket create failed: \(errno)")
            return
        }
        socket = fd

        var opt: Int32 = 1
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &opt, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = Self.broadcastPort.bigEndian
        addr.sin_addr.s_addr = INADDR_ANY

        let bound = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                Darwin.bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bound == 0 else {
            print("[UDP] bind failed: \(errno)")
            stopListening()
            return
        }

        receiveSource = DispatchSource.makeReadSource(fileDescriptor: fd, queue: queue)
        receiveSource?.setEventHandler { [weak self] in
            self?.receive()
        }
        receiveSource?.resume()
        print("[UDP] listening on port \(Self.broadcastPort)")
    }

    private func receive() {
        guard socket >= 0 else { return }
        var buf = [UInt8](repeating: 0, count: 2048)
        var addr = sockaddr_in()
        var len = socklen_t(MemoryLayout<sockaddr_in>.size)
        let n = withUnsafeMutablePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { ptr in
                recvfrom(socket, &buf, buf.count, 0, ptr, &len)
            }
        }
        guard n > 0 else { return }
        let data = Data(bytes: buf, count: n)
        guard let message = String(data: data, encoding: .utf8) else { return }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              json["service"] as? String == "nascab-pro-service",
              json["host"] != nil,
              json["port"] != nil else {
            return
        }
        var serverInfo = json
        withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                var hostBuffer = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                if getnameinfo(sa, socklen_t(MemoryLayout<sockaddr_in>.size), &hostBuffer, socklen_t(hostBuffer.count), nil, 0, NI_NUMERICHOST) == 0 {
                    serverInfo["discovered_ip"] = String(cString: hostBuffer)
                }
            }
        }
        DispatchQueue.main.async { [weak self] in
            self?.onServerDiscovered?(serverInfo)
        }
    }

    func stopListening() {
        receiveSource?.cancel()
        receiveSource = nil
        if socket >= 0 {
            Darwin.close(socket)
            socket = -1
        }
        print("[UDP] stopped")
    }

    deinit {
        stopListening()
    }
}
